patches/0050-gst-libav-add-avvideocompare-element.patch

```diff
@@ -0,0 +1,801 @@
++      GST_VIDEO_INFO_FORMAT (&self->vinfo1) !=
++      GST_VIDEO_INFO_FORMAT (&self->vinfo2) ||
++      GST_VIDEO_INFO_FPS_D (&self->vinfo1) !=
++      GST_VIDEO_INFO_FPS_D (&self->vinfo2) ||
++      GST_VIDEO_INFO_FPS_N (&self->vinfo1) !=
++      GST_VIDEO_INFO_FPS_N (&self->vinfo2))
++    goto input_formats_do_not_match;
++
++  ts = gst_util_uint64_scale_int (self->position * GST_SECOND, self->fps_denom,
++      self->fps_num);
++
++  for (collected = pads->data; collected; collected = g_slist_next (collected)) {
++    GstCollectData *data = (GstCollectData *) collected->data;
++    if (data->pad == self->sinkpad1)
++      buf1 = gst_collect_pads_pop (pads, data);
++    else if (data->pad == self->sinkpad2)
++      buf2 = gst_collect_pads_pop (pads, data);
++  }
++
++  /* compare */
++  if (buf1 && buf2) {
++    AVFrame in1 = { 0, };
++    AVFrame in2 = { 0, };
++    GstVideoFrame frame1, frame2;
++
++    /* set caps if not done yet */
++    if (!gst_pad_has_current_caps (self->srcpad)) {
++      GstCaps *caps;
++      GstSegment segment;
++
++      caps = gst_video_info_to_caps (&self->vinfo1);
++
++      GST_OBJECT_UNLOCK (self);
++      gst_pad_set_caps (self->srcpad, caps);
++      GST_OBJECT_LOCK (self);
++
++      gst_caps_unref (caps);
++      gst_segment_init (&segment, GST_FORMAT_TIME);
++      gst_pad_push_event (self->srcpad, gst_event_new_segment (&segment));
++    }
++
++    if (!gst_video_frame_map (&frame1, &self->vinfo1, buf1, GST_MAP_READ))
++      goto map_failed;
++
++    if (!gst_video_frame_map (&frame2, &self->vinfo2, buf2, GST_MAP_READ)) {
++      gst_video_frame_unmap (&frame1);
++      goto map_failed;
++    }
++
++    _fill_avpicture (self, &in1, &frame1);
++    _fill_avpicture (self, &in2, &frame2);
++
++    if (process_filter_graph (self, &in1, &in2) < 0)
++      GST_WARNING_OBJECT (self, "Could not process filter graph");
++
++    gst_video_frame_unmap (&frame1);
++    gst_video_frame_unmap (&frame2);
++  }
++
++  self->position++;
++
++  GST_OBJECT_UNLOCK (self);
++
++  if (buf2)
++    gst_buffer_unref (buf2);
++
++  if (!buf1) {
++    gst_pad_push_event (self->srcpad, gst_event_new_eos ());
++    return GST_FLOW_EOS;
++  }
++
++  GST_BUFFER_TIMESTAMP (buf1) = ts;
++
++  return gst_pad_push (self->srcpad, buf1);
++
++  /* ERRORS */
++not_negotiated:
++  {
++    GST_OBJECT_UNLOCK (self);
++    GST_ELEMENT_ERROR (self, CORE, NEGOTIATION, (NULL),
++        ("No input format negotiated"));
++    return GST_FLOW_NOT_NEGOTIATED;
++  }
++input_formats_do_not_match:
++  {
++    GstCaps *caps1, *caps2;
++
++    GST_OBJECT_UNLOCK (self);
++    caps1 = gst_pad_get_current_caps (self->sinkpad1);
++    caps2 = gst_pad_get_current_caps (self->sinkpad2);
++    GST_ELEMENT_ERROR (self, CORE, NEGOTIATION, (NULL),
++        ("input formats don't match: %" GST_PTR_FORMAT " vs. %" GST_PTR_FORMAT,
++            caps1, caps2));
++    if (caps1)
++      gst_caps_unref (caps1);
++    if (caps2)
++      gst_caps_unref (caps2);
```

Comment:
You may call `gst_caps_unref()` without the check against NULL.

Reply:
updated